Native Windows support

vi3itor opened the pull request:
Hi @qstanczyk!
I made some changes to the engine and GFootball to be able to run it on Windows natively. It’s been tested on Linux and macOS and even compiles and runs on Apple Silicone, although Python dependencies (opencv, numpy, etc.) must be installed manually for now.

The main changes to the engine are:
- use of vectors instead of variable-length arrays on Windows (MSVC doesn’t support VALs),
- OS-specific file path handling,
- platform-specific clauses in CMakeLists.txt. 

I also ensured that development install (pip install -e .) works well on all supported platforms. Because of a peculiar location of packages in the source tree (`gfootball` and `third_party/gfootball_engine`), `gfootball_engine` has to be copied into the project’s root (as you already doing it on Unix). The only alternative solution I could find requires editing PYTHONPATH, which is prone to errors.   

I have created a document with guides on compiling the engine on different platforms. I will add macOS-related instructions and will go through the issues to compile a list of common problems and solutions.

Windows has a package manager, `vcpkg`, but it’s not as easy to install and use as `brew` on macOS. On the other hand, Windows build collects all the dlls and puts them next to the compiled extension, making it painless to create and distribute wheels. 
I am thinking of creating a GitHub workflow to compile and upload wheels to PyPI automatically.  In such a case, Windows users will only need Python and Git installed on their machines. Will you accept it, or are there some company policies that don’t allow such automation?

I’ll comment on code changes and will ask folks to test it out on their machines. Please let me know if you have any conflicting changes waiting to be published.

vi3itor commented on MANIFEST.in:
```diff
@@ -1,3 +1,11 @@
-include gfootball/build_game_engine.sh
-recursive-include third_party/gfootball_engine/ *
-recursive-include third_party/fonts/ *
+include gfootball/build_game_engine.*
+recursive-include gfootball/doc *
+recursive-include gfootball/eval_server/proto *
+
+recursive-include third_party/gfootball_engine *
+recursive-exclude third_party/gfootball_engine *.a libgame.so cmake_install.cmake Makefile
+prune third_party/gfootball_engine/CMakeFiles
+prune third_party/gfootball_engine/cmake-build-debug
+prune third_party/gfootball_engine/build*
+prune third_party/gfootball_engine/.vs
+prune third_party/gfootball_engine/.idea
```
I reorganized Manifest.in to include non-python files in gfootball and exclude generated files in gfootball_engine.

vi3itor commented on setup.py:
```diff
@@ -16,39 +16,53 @@
 from setuptools.command.install import install
 from setuptools.command.build_ext import build_ext
 
+VERSION = '2.10b3'
```
This line is used for downloading dlls from github releases and will be removed once the wheels are automatically uploaded to PyPI.

vi3itor commented on setup.py:
```diff
@@ -16,39 +16,53 @@
   def __init__(self, name):
     # don't invoke the original build_ext for this special extension
-    super().__init__(name, sources=[])
+    # setuptools is having troubles to build empty list of sources on Windows
+    sources = ['third_party/gfootball_engine/src/misc/empty.cpp'] if platform.system() == 'Windows' else []
+    super().__init__(name, sources=sources, optional=True)
```
Setuptools raises an exception on Windows if the list of sources is empty. Optional argument helps to proceed installation, but it fails in development (editable) mode (pip install -e .). That's why an empty cpp file is necessary to proceed.

qstanczyk replied:
Lets just add empty.cpp unconditionally then

vi3itor commented on setup.py:
```diff
@@ -69,14 +83,81 @@ def run(self):
+    try:
+      download_engine = int(os.environ.get('DOWNLOAD_ENGINE', '0'))
+    except ValueError:
+      raise ValueError('Could not parse COMPILE_ENGINE environment '
```
I used to have COMPILE_ENGINE variable, but changed it to DOWNLOAD_ENGINE. I'll fix it in the next commit.

vi3itor commented on setup.py:
```diff
@@ -69,14 +83,81 @@ def run(self):
+    else:  # compile the engine
+      # help message
+      guide_message = 'Please follow the guide on how to install prerequisites: ' \
+                      'https://github.com/vi3itor/football/blob/windows/gfootball/doc/compile_engine.md#windows'
```
Will point it to the upstream repository in the final commit as well.

vi3itor commented on setup.py:
```diff
@@ -69,14 +83,81 @@ def run(self):
+        shutil.copy2(file, dest_dir)
+
+
+# To support development (a.k.a. editable) install (`pip install -e .` or `python3 setup.py develop`),
+# we remove `build` directory created by a regular setup
+if 'develop' in sys.argv and os.path.exists('build'):
+  shutil.rmtree('build')
```
If a `build` directory was already created, e.g. by running `pip install .` previously, then it will be used by the development install. In such a case, import of `gfootball_engine` will fail, since the the package is not located in the project root.

vi3itor commented on third_party/gfootball_engine/CMakeLists.txt:
```diff
@@ -28,53 +42,67 @@ FIND_PACKAGE(SDL2 REQUIRED)
-FIND_PACKAGE(SDL2_image REQUIRED)
-include_directories(${SDL2_IMAGE_DIRS})
+if(WIN32)
+  FIND_PACKAGE(SDL2-image REQUIRED)
+  FIND_PACKAGE(SDL2-ttf REQUIRED)
+  FIND_PACKAGE(SDL2-gfx REQUIRED)
```
SDL2 packages have different naming on Windows :(

WesleyKhoiVo commented on third_party/gfootball_engine/src/onthepitch/player/humanoid/humanoid.cpp:
```diff
@@ -1847,7 +1847,11 @@ signed int Humanoid::GetBestCheatableAnimID(const DataSet &sortedDataSet, bool u
+#ifdef WIN32
+    std::vector<int> touchIDs(totalTouches);
+#else
     int touchIDs[totalTouches];
```
this too, consider to remove VLA?

WesleyKhoiVo commented on third_party/gfootball_engine/src/onthepitch/teamAIcontroller.cpp:
```diff
@@ -522,7 +522,11 @@ void TeamAIController::CalculateDynamicRoles() {
+#ifdef WIN32
+    std::vector<int> r(playerNum * playerNum);
+#else
     int r[playerNum * playerNum];
```
this too, VLA to vector

WesleyKhoiVo commented on third_party/gfootball_engine/src/onthepitch/teamAIcontroller.cpp:
```diff
@@ -537,7 +541,11 @@ void TeamAIController::CalculateDynamicRoles() {
+#ifdef WIN32
+    int** m = array_to_matrix(&r[0], playerNum, playerNum);
+#else
     int** m = array_to_matrix(r, playerNum, playerNum);
```
VLA to std::vector too

WesleyKhoiVo commented on third_party/gfootball_engine/src/main.hpp:
```diff
@@ -74,10 +74,20 @@ class GameConfig {
   int render_resolution_x = 1280;
   int render_resolution_y = 720;
   std::string updatePath(const std::string& path) {
+#ifdef WIN32
```
Could we consider use same API (C++ 17 filesystem or use boost filesystem) instead split Win32 and Unix file path?

WesleyKhoiVo commented on third_party/gfootball_engine/src/loaders/aseloader.cpp:
```diff
@@ -395,7 +398,11 @@ void ASELoader::BuildTriangleMesh(
+#ifdef WIN32
+    std::vector<Vector3> tvertex_cache(numtvertex);
+#else
     Vector3 tvertex_cache[numtvertex];
```
VLA to std::vector too

WesleyKhoiVo commented on third_party/gfootball_engine/src/data/teamdata.cpp:
```diff
@@ -198,7 +198,11 @@ TeamData::TeamData(int teamDatabaseID, const std::vector<FormationEntry> &f) {
+#ifdef WIN32
+    std::vector<Vector3> offset(player_count);
+#else
     Vector3 offset[player_count];
```
Hi @vi3itor, do you consider remove Variable Length Arrays in next commit? I think it will be better use std::vector for both MSVC and Clang.

WesleyKhoiVo commented on third_party/gfootball_engine/src/loaders/aseloader.cpp:
```diff
@@ -222,8 +222,11 @@ void ASELoader::BuildTriangleMesh(
+#ifdef WIN32
+  std::vector<Vector3> vertex_cache(numvertex);
+#else
   Vector3 vertex_cache[numvertex];
```
VLA to std::vector too

WesleyKhoiVo commented on gfootball/doc/compile_engine.md:
```diff
@@ -0,0 +1,112 @@
+Prerequisites:
+- [Git](https://git-scm.com/download/win),
+- [CMake](https://cmake.org/download/),
+- [Visual Studio 2019 Community Edition](https://visualstudio.microsoft.com/downloads/) make sure to 
```
I was used VS 2017 to build, not only require to use VS 2019, but may be we should consider using newest IDE to keep away using boost

WesleyKhoiVo reviewed (commented):
First quick review

vi3itor commented:
@WesleyKhoiVo At first, I replaced all VLAs with std::vector, and of course it worked. But, as I mentioned above, since I cannot reliably test if there is any performance drop, I decided to do it only for Windows, and leave Unix version as it was before. We can easily change it to std::vector in the future, if the engine will undergo a significant refactoring.

Same logic goes for filesystem calls: since the engine was optimized for speed, I would leave it as it is on Unix, and change behaviour only on Windows. Although I don't think there is any significant overhead.

As for the third point, I recommend installing VS2019 because it's the latest version, but I agree that it should compile on VS2017 or even earlier.

WesleyKhoiVo commented:
Which OS drop performance? Unix or Windows? Visual Studio include profiling tool, I will try to debug it to find out why

qstanczyk reviewed (commented):
Thanks Victor for the change, it is grate. Regarding deployment of compiled binaries - it would be great if you could set that up on GitHub. We don't have it automated for this project. Also, there are no in flight changes that would conflict with this work, so feel free to ping me when ready to merge the change.

WesleyKhoiVo reviewed (commented):
No need convert array to vector, due to constant not variable

vi3itor commented:
@qstanczyk Hi Piotr! 
I finished with the PR. Now it includes a GitHub workflow that **on publishing the release** will build eight wheels for Windows (Python 3.6-3.9 both x86 and x64), run the tests, and, in case of success, upload them to the PyPI together with the source code. I wanted to set up continuous integration (and possibly wheel generation) for other platforms, including on pull requests, but it is better to do it one mile at a time. I will prepare other PRs later with the CI and updated build instructions for the remaining platforms.

To automate uploading the wheels and source code, you need to create API token on PyPI and an associated GitHub secret `PYPI_API_TOKEN` for the repository. Please refer to the description [here](https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/). There is no need to create TestPyPI token.

As for some minor changes since the previous commits that I'd like to clarify: 
* I have updated `expected_hash_values` to pass rendering tests. The values differ by platform, but there are some of the checks that seem outdated, see `football_env_test.py` lines 150, 159, 168. Feel free to remove them or to leave them as is.
* I added `requirements.txt` to install GRF from the built wheel. It adds extra work to keep the list of required packages here and in the `setup.py`, but it provides greater flexibility.

This PR and previous commits solve some existing issues (#214 #238 #249 #247 #241 #233 #136). I think they can be closed now.

qstanczyk commented:
Thanks for all the changes. I have created an API token, so feel free to play around with it. It would be great if we could automate releases for all platforms.

vi3itor commented:
@qstanczyk
Yes, we'll automate builds for all platforms. Continuous Integration is the first step, followed by distributing binary wheels for macOS and Linux with the help of [`delocate`](https://github.com/matthew-brett/delocate) and [`auditwheel`](https://github.com/pypa/auditwheel).

But to publish the new version of `gfootball` that supports Windows to PyPI now, please don't forget to create a new release by going to the [corresponding page](https://github.com/google-research/football/releases/new), creating a new tag `v2.10` and clicking `Publish release` button. It will automatically run the workflow that builds wheels for Windows, runs the tests, and, in case of success, uploads wheels and the zipped source code (`sdist`) to the PyPI. From now on no more more manual uploading for the future releases: updating version string in the `setup.py` and publishing the release through the GitHub interface will suffice.

And what about closing some of the existing issues that I pointed [above](https://github.com/google-research/football/pull/277#issuecomment-918319288)? Do you need more time or details, or you simply haven't noticed the suggestion?

qstanczyk commented:
Thanks Victor, closed the issues you pointed at.
Also, triggered 2.10 release, but building failed with (can you investigate it? Thanks a lot):

-- Downloading https://github.com/python/cpython/archive/v3.6.4.tar.gz -> python-cpython-v3.6.4.tar.gz...
-- Extracting source D:/a/football/football/vcpkg/downloads/python-cpython-v3.6.4.tar.gz
-- Applying patch D:/a/football/football/vcpkg/buildtrees/versioning/versions/python3/31048e50e35696fa4eaa3cecf873a2cc96cb58fb/0004-Fix-iomodule-for-RS4-SDK.patch
-- Applying patch D:/a/football/football/vcpkg/buildtrees/versioning/versions/python3/31048e50e35696fa4eaa3cecf873a2cc96cb58fb/0005-Fix-DefaultWindowsSDKVersion.patch
-- Applying patch dev16.patch
-- Applying patch Microsoft.VisualStudio.Setup.Configuration.Native.patch
-- Using source at D:/a/football/football/vcpkg/buildtrees/python3/src/v3.6.4-e4fea3fdbc.clean
-- Building D:/a/football/football/vcpkg/buildtrees/python3/src/v3.6.4-e4fea3fdbc.clean-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj for Release
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: msbuild D:/a/football/football/vcpkg/buildtrees/python3/src/v3.6.4-e4fea3fdbc.clean-Lib-dynamic-crt-dynamic/PCBuild/pythoncore.vcxproj /p:Configuration=Release /t:Rebuild /p:Platform=x64 /p:PlatformToolset=v142 /p:VCPkgLocalAppDataDisabled=true /p:UseIntelMKL=No /p:WindowsTargetPlatformVersion=10.0.22000.0 /p:VcpkgManifestInstall=false /p:VcpkgManifestEnabled=false /m /p:VcpkgEnabled=false
    Working Directory: D:/a/football/football/vcpkg/buildtrees/python3/x64-windows-rel
    Error code: 1
    See logs for more information:
      D:\a\football\football\vcpkg\buildtrees\python3\build-x64-windows-rel-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_msbuild.cmake:129 (vcpkg_execute_required_process)
  buildtrees/versioning/versions/python3/31048e50e35696fa4eaa3cecf873a2cc96cb58fb/portfile.cmake:57 (vcpkg_build_msbuild)
  scripts/ports.cmake:140 (include)


Error: Building package python3:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: python3:x64-windows
  Vcpkg version: 2021-08-12-85ab112d5ee102bc6eac8cdbbfdd173a71374e04

Additionally, attach any relevant sections from the log files above.
-- Running vcpkg install - failed
CMake Error at D:/a/football/football/vcpkg/scripts/buildsystems/vcpkg.cmake:827 (message):
  vcpkg install failed.  See logs for more information:
  C:\Users\runneradmin\AppData\Local\Temp\build-via-sdist-4okp6qo1\gfootball-2.10\third_party\gfootball_engine\build_win\vcpkg-manifest-install.log
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.21/Modules/CMakeDetermineSystem.cmake:124 (include)
  CMakeLists.txt:28 (project)


-- Configuring incomplete, errors occurred!
error: Google Research Football compilation failed.
Please follow the guide on how to install prerequisites: https://github.com/google-research/football/blob/master/gfootball/doc/compile_engine.md#windows

ERROR Backend subproccess exited when trying to invoke build_wheel
Error: Process completed with exit code 1.

vi3itor commented:
@qstanczyk 
I'm investigating it now. Looks like there were some changes to the default software (like Visual Studio or CMake) in the virtual machine in the past 9 days or so. As you can see some [previous runs in the temporary repository](https://github.com/vi3itor/temp-pypi-gfootball/actions) it worked well as recent as last week.

I will try to fix it asap and get back to you tomorrow.

qstanczyk commented:
Thanks a lot.

vi3itor commented:
I think I found the cause of the problem: Windows 11 Preview SDK (10.0.22000.0) was added recently to the [GitHub's virtual environment](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md) and causing troubles to build Python. There is even a corresponding issue on Python's bug tracker [posted 7 days ago](https://bugs.python.org/issue45220), that prevented CI for Python builds. 
It seems that the issue affects not only Python language development team, but also any `vcpkg` CI pipeline that builds Python from sources. I tried creating the simplest job that runs `vcpkg install python3` and [it fails](https://github.com/vi3itor/gh-workflows-test/actions). I will open the corresponding issues in the `vcpkg` and `virtual-environment` repositories, and report to Windows SDK team.  

As for the Windows users of the `gfootball`, unless someone installs buggy Windows 11 Preview SDK, they shouldn't run into this problem. I couldn't reproduce it on the local Windows machines, even after installing exact versions of Visual Studio, CMake and other software as in the GitHub's virtual environment.

But for building the wheels on GitHub's CI and publishing them to PyPI automatically, I'm still looking for a quick fix. We can always fork `vcpkg` repository and apply patches to Python sources to ignore the above mentioned version of SDK (as it is currently done by Python team), but I'm looking for a less hacky solution, such as defining environment variables. I'll get back to you as soon as the fix is ready.

vi3itor commented:
Well, the issue [has been already reported](https://github.com/microsoft/vcpkg/pull/20292) to `vcpkg` and the PR with patch is ready. As it is pointed out, the problem was introduced with the [PR to the `virtual-environments`](https://github.com/actions/virtual-environments/pull/4014) and by default `vcpkg` gets the latest WinSDK available in the system.

qstanczyk commented:
I guess we can wait for the fix to hit production then. WDYT?

vi3itor commented:
I'm afraid that it might take some time for the proper fix to arrive. In the meantime I think we should run the build on `windows-2016` machines. It has Visual Studio 2017 installed (instead of 2019) and it uses its SDKs. I successfully [built the wheels](https://github.com/vi3itor/temp-pypi-gfootball/actions/runs/1266644202) and tested them on the latest version of Windows. Everything works well, although the hashes for the rendering tests had to be updated. I will send a PR shortly.
